sys/fork.c

```diff
@@ -0,0 +1,57 @@
+#include <fork.h>
+#include <thread.h>
+#include <sync.h>
+#include <filedesc.h>
+#include <sched.h>
+#include <stdc.h>
+
+int do_fork() {
+  thread_t *td = thread_self();
+
+  /* Cannot fork non-user threads. */
+  assert(td->td_uspace);
+
+  thread_t *newtd = thread_create(td->td_name, NULL, NULL);
+
+  /* Clone the thread. Since we don't use fork-oriented thread_t layout, we copy
+     all fields one-by one for clarity. The new thread is already on the
+     all_thread list, has name and tid set. */
+
+  newtd->td_state = td->td_state;
+  newtd->td_flags = td->td_flags;
+  newtd->td_csnest = td->td_csnest;
+
+  /* Copy user context.. */
+  newtd->td_uctx = td->td_uctx;
+  newtd->td_uctx_fpu = td->td_uctx_fpu;
+  exc_frame_set_retval(&newtd->td_uctx, 0);
+
+  /* New thread does not need the exception frame just yet. */
+  newtd->td_kframe = NULL;
+  newtd->td_onfault = 0;
+
+  /* The new thread already has a new kernel stack allocated. There is no need
+     to copy its contents, it will be discarded anyway. We just prepare the
+     thread's kernel context to a fresh one so that it will continue execution
+     starting from user_exc_leave (which serves as fork_trampoline). */
+  ctx_init(newtd, user_exc_leave, NULL);
+
+  /* Share userspace memory with parent. TODO: (optionally?) clone user
+     memory. */
+  newtd->td_uspace = td->td_uspace;
```

Comment:
Actually... I don't know how to copy memory contents. Only one `vm_map` is active at a time, so I would need to copy out contents somewhere (where? kernel memory? how much space?), and switch between `vm_map`s back and forth... which seems like a very primitive solution.

I suppose no reference OS implements such mechanism (they all prefer COW + a hierarchy of shadow vm objects etc.), so I don't think I can find help there.

The best idea I have so far is to create a new pager type, which clones pages on demand. It would have to serve both the parent and child process, to avoid inconsistencies.

Or maybe this is the time we've started working on a simple layered hierarchy of vm_objects?